library/core/class.format.php

```diff
@@ -1430,6 +1430,52 @@ public static function Mentions($Mixed) {
       }
    }
 
+   /** Do a preg_replace, but don't affect things inside <code> tags.
+    * The three parameters are identical to the ones you'd pass
+    * preg_replace.
+    *
+    * @param mixed $Pattern
+    * @param mixed $Replacement
+    * @param mixed $Subject
+    * @param bool $IsCallback If true, do preg_replace_callback. Do
+    *             preg_replace otherwise.
+    * @return string
+    */
+   public static function ReplaceButProtectCodeBlocks($Pattern, $Replacement, $Subject, $IsCallback = FALSE) {
+     // Take the code blocks out, replace with something unlikely to
+     // appear in the string, and keep track of what substring got
+     // replaced with what code.
+     $Replacements = array();
+     $Subject = preg_replace_callback(
+       '/<code>.*?<\/code>/i',
+       function($Matches) use (&$Replacements) {
+         // Random string and replacement index, surrounded by
+         // whitespace to try to prevent the characters from being
+         // picked up by $Pattern.
+         $ReplacementString = ' 5z9Ah9Y2sX5xnzUx8wSq'.count($Replacements).' ';
+         $Replacements[$ReplacementString] = $Matches[0];
+         return $ReplacementString;
+       },
+       $Subject
+     );
+
+
+     // Do the requested replacement.
+     if ($IsCallback) {
+       $Subject = preg_replace_callback($Pattern, $Replacement, $Subject);
+     } else {
+       $Subject = preg_replace($Pattern, $Replacement, $Subject);
+     }
+
+     // Put back the code blocks. First in, last out so that e.g.
+     // random10 is not replaced by random1.
+     foreach (array_reverse($Replacements) as $Placeholder => $Original) {
+       $Subject = str_replace($Placeholder, $Original, $Subject);
```

Comment:
[str_replace](http://php.net/str_replace) supports arrays in $search and $replace.  If you kept your placeholders in one array and your code block content in another, so long as they're synchronized, you should be able to ditch this foreach loop.